D1: Use array-based runtime handlers insead of variadic version

mihails-strasuns-sociomantic opened the pull request:
Essentially is a backport of cab160591b0da9a7249b43a62319e0862e292eb9 with
tweaks to dmd1 state of code.

Tested locally and ready to review/merge. Does require matching runtime
(https://github.com/sociomantic-tsunami/tangort) modification, but that can/will be
handled on our side.

This is one of pre-requisites I have found for https://github.com/dlang/dmd/pull/6642

mihails-strasuns-sociomantic committed:
Use array-based NEWARRAYM*T insead of variadic version

Backport of cab160591b0da9a7249b43a62319e0862e292eb9

mihails-strasuns-sociomantic commented:
FYI @leandro-lucarella-sociomantic @WalterBright

mihails-strasuns-sociomantic committed:
Use array-based ARRAYCATN* instead of variadic version

Backport of 599a312b7b2ea0af30b70edd748e70096aa2ae15

mihails-strasuns-sociomantic commented:
Added second commit which does the same for `_d_arraycatnTX`

WalterBright commented on src/e2ir.c:
```diff
@@ -1943,25 +1952,32 @@ elem *CatExp::toElem(IRState *irs)
     {
-        elem *ep;
         CatExp *ce = this;
         int n = 2;
```
I think you can delete `n`.

WalterBright commented:
Your strategy of keeping it line-by-line with the D2 version of e2ir.d is a good one. I see you left off the Win64 support, which is reasonable if you've got no plans to ever support that.

mihails-strasuns-sociomantic commented:
Updated:
- removed unused `n` variable

> I see you left off the Win64 support, which is reasonable if you've got no plans to ever support that.

Yes, we don't use/test Windows builds at all in our projects and AFAIK dmd1 had never supported Win64 anyway, so it is better to not pretend otherwise :)

WalterBright commented:
Isn't it ironic that `e2ir.c` was translated to D, and now back to C++!

mihails-strasuns-sociomantic commented:
@WalterBright does it look good to merge from your side?

WalterBright commented:
yes
